File: leetcode/golang/LCR_153_er_cha_shu_zhong_he_wei_mou_yi_zhi_de_lu_jing_lcof.go
type TreeNode struct {
        Val int
        Left *TreeNode
        Right *TreeNode
}

func pathTarget(root *TreeNode, target int) [][]int {
    if root == nil {
        return [][]int{}
    }
    var res [][]int
    var tmp []int
    var find  func(root *TreeNode, target int) 
    find = func(root *TreeNode, target int) {
        if root == nil {
            return
        }
        tmp = append(tmp, root.Val)
        if root.Left == nil && root.Right == nil && root.Val == target {
            res = append(res, append([]int(nil), tmp...))
        } else {
            find(root.Left, target-root.Val)
            find(root.Right, target-root.Val)
        }
        tmp = tmp[:len(tmp)-1]
    }
    find(root, target)
    return res
}

// res = append(res, append([]int(nil), tmp...))这个是关键，直接添加tmp的话，相当于只添加了tmp对象，后面改就会更改，所以需要添加tmp的复制

File: leetcode/c++/LCR_153_er_cha_shu_zhong_he_wei_mou_yi_zhi_de_lu_jing_lcof.cpp
// https://leetcode.cn/problems/er-cha-shu-zhong-he-wei-mou-yi-zhi-de-lu-jing-lcof/description

#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
      int val;
      TreeNode *left;
      TreeNode *right;
      TreeNode() : val(0), left(nullptr), right(nullptr) {}
      TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
      TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    void find(TreeNode* root, int target, vector<vector<int>>& res, vector<int>& tmp) {
        if (root == nullptr) 
            return;
        tmp.push_back(root->val);
        if (root->left == nullptr && root->right == nullptr && target == root->val) {
            res.push_back(tmp);
        } else {
            find(root->left, target-root->val, res, tmp);
            find(root->right, target-root->val, res, tmp);
        }
        tmp.pop_back();
    }
    vector<vector<int>> pathTarget(TreeNode* root, int target) {
        vector<vector<int>> res;
        vector<int> tmp;
        if (root == nullptr) 
            return res;
        find(root, target, res, tmp);
        return res;
    }
};

// 时间复杂度为O(n), 空间复杂度为O(n)